File: leetcode/medium/4Sum.py
from typing import List

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        """
        Finds all unique quadruplets in the array which give the sum of target.

        This solution follows a generalized approach for k-Sum problems. It uses
        sorting and a two-pointer technique to achieve an O(n^3) time complexity.

        Algorithm Breakdown:
        1. Sort the input array. This is crucial for using the two-pointer
           technique and for efficiently skipping duplicate numbers.
        2. Iterate through the array with a primary pointer `i`. This will be the
           first number in our potential quadruplet.
        3. For each `i`, iterate with a second pointer `j` starting from `i + 1`.
           This will be the second number.
        4. For each pair `(i, j)`, we now need to solve a 2-Sum problem for the
           remaining part of the array. The new target will be `target - nums[i] - nums[j]`.
        5. Use two pointers, `left` and `right`, to scan the subarray from `j + 1`
           to the end. `left` starts at `j + 1` and `right` starts at the end of the array.
        6. Move the `left` and `right` pointers inwards based on their sum:
           - If `sum < target`, move `left` right to increase the sum.
           - If `sum > target`, move `right` left to decrease the sum.
           - If `sum == target`, we've found a quadruplet.
        7. To ensure uniqueness, skip duplicate numbers at every stage (for `i`,
           `j`, `left`, and `right`) after they have been processed.

        Time Complexity: O(n^3), due to the nested loops and the two-pointer scan.
        Space Complexity: O(1) if we don't count the space for the result list.
        """
        n = len(nums)
        if n < 4:
            return []

        # 1. Sort the array
        nums.sort()
        result = []

        # 2. Outer loop for the first number
        for i in range(n - 3):
            # Skip duplicates for the first number
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            # --- Early exit optimizations ---
            # If the smallest possible sum from this point is > target, we can stop.
            if nums[i] + nums[i+1] + nums[i+2] + nums[i+3] > target:
                break
            # If the largest possible sum from this point is < target, we can skip to the next i.
            if nums[i] + nums[n-1] + nums[n-2] + nums[n-3] < target:
                continue

            # 3. Inner loop for the second number
            for j in range(i + 1, n - 2):
                # Skip duplicates for the second number
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                
                # --- Early exit optimizations for the inner loop ---
                if nums[i] + nums[j] + nums[j+1] + nums[j+2] > target:
                    break
                if nums[i] + nums[j] + nums[n-1] + nums[n-2] < target:
                    continue

                # 5. Two-pointer approach for the remaining two numbers
                left, right = j + 1, n - 1
                while left < right:
                    current_sum = nums[i] + nums[j] + nums[left] + nums[right]

                    if current_sum < target:
                        left += 1
                    elif current_sum > target:
                        right -= 1
                    else:
                        # Found a quadruplet
                        result.append([nums[i], nums[j], nums[left], nums[right]])
                        
                        # 7. Skip duplicates for the third and fourth numbers
                        # Move left pointer forward while the value is the same
                        left += 1
                        while left < right and nums[left] == nums[left - 1]:
                            left += 1
                        
                        # Move right pointer backward while the value is the same
                        right -= 1
                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1
                            
        return result

File: leetcode/medium/4Sum.java

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Solution for the 4Sum problem.
 * The method finds all unique quadruplets in the array which give the sum of the target.
 * The overall time complexity is O(n^3) due to the nested loop structure.
 * The space complexity is O(1) for the variables, and the space for the result list is not counted.
 */
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        // Sort the array to enable the two-pointer approach and easily handle duplicates.
        Arrays.sort(nums);
        int length = nums.length;
        List<List<Integer>> res = new ArrayList<>();

        // Iterate through the array, considering each element as a potential first element of a quadruplet.
        for(int i = 0; i < length - 3; i++) {
            // Skip duplicate elements for the first element to avoid duplicate quadruplets.
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            // Iterate through the array, considering each element as a potential second element of a quadruplet.
            for (int j = i + 1; j < length - 2; j++) {
                // Skip duplicate elements for the second element to avoid duplicate quadruplets.
                if (j > i + 1 && nums[j] == nums[j - 1]) {
                    continue;
                }

                // Use two pointers for the remaining two elements.
                int left = j + 1, right = length - 1;
                while (left < right) {
                    // Using long for total to prevent integer overflow
                    long total = (long) nums[i] + nums[j] + nums[left] + nums[right];

                    // If the sum is less than the target, we need a larger sum, so move the left pointer.
                    if (total < target) {
                        left++;
                    }
                    // If the sum is greater than the target, we need a smaller sum, so move the right pointer.
                    else if (total > target) {
                        right--;
                    }
                    // If the sum is equal to the target, we found a quadruplet.
                    else {
                        res.add(List.of(nums[i], nums[j], nums[left], nums[right]));
                        
                        // Skip duplicate elements for the third element.
                        while (left < right && nums[left] == nums[left + 1]) {
                            left++;
                        }

                        // Skip duplicate elements for the fourth element.
                        while (left < right && nums[right] == nums[right - 1]) {
                            right--;
                        }

                        // Move both pointers to find the next potential quadruplet.
                        left++;
                        right--;
                    }
                }

            }
        }
        return res;
    }
}